### PR TITLE
Election setup wizard (Iteration 1) 

### LINKED ISSUE #56: Election setup wizard (Iteration 1)
Ref: VOC-SEC-03

As an administrator, to create a new election I have to perform a number of actions (enter the basic information, define the questions and answers, upload the voter census) before setting it up in the Vocdoni API.

Currently to do this I have the actions separately and a [dashboard](https://github.com/decidim-vocdoni/decidim-module-vocdoni/issues/9) that shows me the pending steps. To improve the flow of creating an election, I want to have a wizard that guides me through the whole process.

The wizard will have the following sections:

1. Basic information
    * Title
    * Description
    * Image
    * Video stream link
2. Questions
3. Census
4. Calendar and results
5. Publish

#### Admin panel -  Elections
imatge

#### Wizard

Tab #1 Basic info

Tab #2 Questions

Tab #3 Census page

Tab #4 Calendar and results

Tab #5 Publish

#### Dashboard

dashboard-setup

### Acceptance criteria

- [ ] Given that I'm an administrator,
    When I start creating a new election,
    Then I am presented with a wizard that guides me through the following sections:
    
    Basic Information:
    - [ ] I can enter the title, description, image, and video stream link for the election.
    - [ ] I must provide all the required information before proceeding to the next section.

    Questions:
    - [ ] I can define the questions and answers for the election.
    - [ ] I can add multiple questions and provide answer options for each question.
    - [ ] I must provide at least one question before proceeding to the next section.

    Census:
    - [ ] I can upload the voter census for the election.
    - [ ] I have the option to upload a file or use the decidim permissions system.
    - [ ] I must upload the census or assign some permissions before proceeding to the next section.

    Calendar and Results:
    - [ ] I can set the election calendar, including the start and end dates.
    - [ ] I can specify when the results will be announced.
    - [ ] I must provide the election dates and results announcement date before proceeding to the next section.

    Publish:
    - [ ] I have the option to go back and edit any section if needed.
    - [ ] I can choose to publish the election once I'm satisfied with the information.

**Additional Acceptance Criteria:**

- [ ] Given that I'm an administrator,
    When I navigate to the dashboard,
    Then I see a clear indication of the progress and pending steps for each election being created.
- [ ] Given that I'm an administrator,
    When I edit an existing election in the wizard,
    Then I can modify any section and update the information accordingly.
- [ ] Given that I'm an administrator,
    When I cancel the creation of an election in the wizard,
    Then any entered information is discarded, and I am returned to the dashboard.
- [ ] Given that I'm an administrator,
    When I encounter any errors or missing information while progressing through the wizard,
    Then I receive clear and informative error messages indicating the required actions to proceed.

### PR DESCRIPTION
**Pull request by antopalidi**
fixes #56 

In this PR the election wizard is implemented by using navigation between tabs

Election creation:

1. New election on the index elections page
Screenshot 2023-05-26 at 11 20 31

2. After filling in the required fields, you can go to the next step. This step creates election.  If the required fields are not filled in, form errors will be displayed.
Screenshot 2023-05-26 at 11 21 41

3. Adding questions and answers. Go to the next step after a minimum of 1 question and each question has a minimum of 2 answers.
Screenshot 2023-05-26 at 11 24 04
Screenshot 2023-05-29 at 13 43 52

4. Census. After uploading the census, the election calendar and other settings can be added.
Screenshot 2023-05-26 at 11 25 03

5. Calendar. After filling in the required fields of the form in the election calendar, you can go to Publishing. If the required fields are not filled in, form errors will be displayed.
Screenshot 2023-05-26 at 11 30 18

7. Publish Elections. After publishing, the setup wizard redirects the user to the steps dashboard. 
Screenshot 2023-05-26 at 11 33 21
Screenshot 2023-05-26 at 11 33 21

### CODE REVIEW

File: app/commands/decidim/vocdoni/admin/create_election.rb
Comment (microstudi) [inline app/commands/decidim/vocdoni/admin/create_election.rb:61]: this attributes are still in the step of creating he election? I think all these steps go in the last step

**Review by microstudi — COMMENTED** [review]
Looks good! it is headed in the right direction!

Some preliminary notes:
We need to mark the active tab to avoid loosing context:

image

When creating/editing a question the layout is lost (no tabs):
image

The main title is incosistent (it says "new election", sometimes only, it should say new election only when creating.
image

**Comment by andreslucena** [comment]
@microstudi can you please give it the first review, and I'll do it after you? Thanks

**Review by microstudi — CHANGES_REQUESTED** [review]
Hi @antopalidi,
this looks awesome. Just a copule of UI fixes to let it perfect:

I've found some places where the tab navigaton is lost:

Editing a question:
image

Editing an answer:
image

Also, when editing/creating answers, we should have the reference to the question we are editing, we could just add the name of the question in the title:
image

**Comment by antopalidi** [comment]
@andreslucena Ivan did the first review, it's done

**Comment by carolromero** [comment]
@antopalidi this works great, nice job!

A couple of things I've seen during the review:
* Once I am in the Steps Dashboard and see what settings are pending, my expectation is that clicking on "Fix it" will take me directly to the appropriate tab. It works perfectly in all cases except Publish, which takes me to the first tab. Can we change it to take you directly to the tab?
https://www.awesomescreenshot.com/video/18245430?key=b47f1935488a35f6e1c699701c5ba2f9
* When adding an image to the election, it apparently does not save it, but it is displayed when published. I mention this since I've not been able to reproduce it another instance where we have this module installed.

**Comment by antopalidi** [comment]
> * When adding an image to the election, it apparently does not save it, but it is displayed when published. I mention this since I've not been able to reproduce it another instance where we have this module installed.

@carolromero Could you clarify this, where the image is not saved?

**Comment by carolromero** [comment]
@antopalidi I have retested it several times now and it works as expected. My bad, I must have mistaken it while testing this morning, sorry!

**Comment by microstudi** [comment]
Deployed again in case of re-testing

**Review by microstudi — CHANGES_REQUESTED** [review]
@antopalidi we still need to indicate the question we are adding or editing an answer for:
image

Currently, when editing/creating an answer you loose reference of which question are you doing it for. Can you add the name of the question? something like:

**Edit answer for question "Hello world?"**
**New answer for question "Hello world?"**

File: app/forms/decidim/vocdoni/admin/setup_form.rb
Comment (microstudi) [inline app/forms/decidim/vocdoni/admin/setup_form.rb:35]: we can remove this

**Review by microstudi — CHANGES_REQUESTED** [review]
@antopalidi all good, just a note please!

File: lib/decidim/vocdoni/admin_engine.rb
Comment (andreslucena) [inline lib/decidim/vocdoni/admin_engine.rb:28]: There's a weird thing with this URL that we should check out. I'm not sure how to actually solve it though. 

If you see how the URLs are now, we have: 

* /elections/new
* /elections/4/edit
* /elections/4/questions
* /elections/4/questions/new
* /elections/4/census
* /elections/4/election_calendar/4/edit
* /elections/4/publish_page

IMHO all make sense except the calendar page, as we have the ID two times (and as far as I see the second time is ignored): `/elections/4/election_calendar/4/edit`. Can we find a way of leaving out only `/elections/4/calendar`?

File: lib/decidim/vocdoni/admin_engine.rb
Comment (andreslucena) [inline lib/decidim/vocdoni/admin_engine.rb:16]: I understand that we can't have it like 

```suggestion
            get :publish
```

as that would generate a conflict with the next line (`put :publish`), right? I say it because it's cleaner, to have it as /publish instead of /publish_page (at least in the URL).

File: config/locales/en.yml
Comment (andreslucena) [inline config/locales/en.yml:144]: I have a couple suggestions here:
* If the string has html, we should add the suffix "_html" so we don't need to call to the html_safe method on the view
* Also I think it's cleaner and more semantic to use the paragraph tag (`<p>`) instead of the break line (`<br>`)
* Finally, if rubocop let us, I prefer to break the 80 characters column convention if we improve the legibility.

This is what I have in mind:  
```suggestion
            publish_text_html: |-
              <p>If you are happy with your configuration (you can still edit all the steps), now it's time to publish and setup de election.</p>
               <p>To setup the election you must publish it first.</p>
               <p>The setup process is when the election configuration, including the census, will be sent to the Vocdoni platform. This means it will be unmodifiable
                after that, due the use of the blockchain. After the setup, the election can start (either manually or automatically depending on the configuration).</p>

```

We would need to change it also on `publish_text_all_done` and also on the view (the `_html` suffix in the variables on both of these and removing the `html_safe` method call).

File: config/locales/en.yml
Comment (andreslucena) [inline config/locales/en.yml:98]: Oh I see that I left the date of birth here, nice catch!

File: app/views/decidim/vocdoni/admin/shared/_election_title.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/shared/_election_title.html.erb:1]: Can you add the HTML title here? 

```suggestion
<% add_decidim_page_title(translated_attribute(election.title)) %>

<div class="card-divider mb-s">
```

File: app/views/decidim/vocdoni/admin/answers/new.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/answers/new.html.erb:1]: On this case and in almost all the pages that I've reviewed, as we're removing this helper we're losing the HTML page titles. Can you add it again on this and all the pages please?

File: app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb
Comment (andreslucena) [inline app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb:38]: As in the other cases, I think that if we change this key name with the `_html` suffix we can remove the `html_safe` from the call.
```suggestion
          t("for_question", question: link, scope: "decidim.vocdoni.admin.answers.index")
```


More information on  https://guides.rubyonrails.org/i18n.html#using-safe-html-translations

File: app/helpers/decidim/vocdoni/admin/steps_helper.rb
Comment (andreslucena) [inline app/helpers/decidim/vocdoni/admin/steps_helper.rb:26]: I'm not familiar with the `concat` method, but maybe we could get away with a bit of string interpolation here? 
```suggestion
            "#{icon(icon)}  #{I18n.t("decidim.vocdoni.admin.steps.create_election.errors.fix_it_text")}"
```

File: app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb
Comment (andreslucena) [inline app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb:25]: We can probably remove the `Metrics/CyclomaticComplexity` cop exception if we refactor this by extracting some of the logic to private methods, for instance: 

```suggestion
            is_publishable? ? "" : "disabled"
```

**Review by andreslucena — CHANGES_REQUESTED** [review]
I've tried it locally and I really like the new flow, it's an improvement to the messy flow that we had until now. Congratulations! 

I have some suggestions on how to improve the code. 

I've also found a bug regarding the permissions of the steps: if you know the URLs you can access to any of these steps without actually checking if the previous steps were fulfilled. 

Steps to reproduce: 

1. Create a new election (on /elections/new)
2. Fill title, description and click on the "Save and go to the next step" button 
3. In the URL, change "/questions" to "/publish_page" 
4. You're on the last step that's publishable, but it's on a state that should actually work (for instance if you go to the frontend page, you have an exception related to the date that's not in the Election)

Screenshot of an election without questions, census, nor calendar and being publicable

Related to this, I've also found a bug with the "publish" icon in the Manage elections table:

1. Create a new election (on /elections/new)
2. Fill title, description and click on the "Save and go to the next step" button 
3. Go to the "Manage elections" table
4. Click on the "Publish" icon for the election
5. Again, it's on a state that should actually work (for instance if you go to the frontend page, you have an exception related to the date that's not in the Election)

Screenshot of the last step done on an election that's not finished

I think we should fix both of these bugs, the first one by checking if the election has actually fulfilled the previous step (on the controller level) and the second one by reusing part of the logic of the `tab_link_class` method helper so it's actually checked on the `PublishElection` command. Can you handle that 🙏🏽 ? Thanks

File: lib/decidim/vocdoni/admin_engine.rb
Comment (antopalidi) [reply lib/decidim/vocdoni/admin_engine.rb:16]: @andreslucena I didn't get this point very well. We have two actions in the controller:

`publish_page` is the get action showing the html
 `publish` is the put action to publish the election

We need two methods, just like new and create. I can't make the URL for publish_page the same as for publish. Maybe it would be preferable to change the URL? For example publish_election instead of publish page

Comment (andreslucena) [reply lib/decidim/vocdoni/admin_engine.rb:16]: I think you understood me perfectly! The create/new example is spot on. 

>  Maybe it would be preferable to change the URL? For example publish_election instead of publish page

Yes, that would be great. `publish_page` I think it could introduce some confusion (there's a module `decidim-pages` that has not relation with this feature), and `publish_election` is more clear. Can you change that please?

File: app/views/decidim/vocdoni/admin/answers/edit.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/answers/edit.html.erb:1]: Can we leave the resource title in the page title please? So it's easier to differentiate from other resources.  Also let's have a break-line (as it was before) so it easy to have the meta and body blocks visually separated. 

```suggestion
<% add_decidim_page_title(t(".title")) %>
<% add_decidim_page_title(translated_attribute(answer.title)) %>

```

File: app/views/decidim/vocdoni/admin/answers/index.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/answers/index.html.erb:1]: Let's introduce a break-line (as it was before) so it easy to have the meta and body blocks visually separated. 

```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/answers/new.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/answers/new.html.erb:1]: Let' add a break-line (as it was before) so it easy to have the meta and body blocks visually separated. 

```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/census/index.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/census/index.html.erb:1]: Can we leave the resource title in the page title please? So it's easier to differentiate from other resources.  Also let's add a break-line (as it was before) so it easy to have the meta and body blocks visually separated. 

```suggestion
<% add_decidim_page_title(translated_attribute(election.title)) %>
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/election_calendar/edit.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/election_calendar/edit.html.erb:1]: Let' add a break-line (as it was before) so it easy to have the meta and body blocks visually separated.
```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/elections/edit.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/elections/edit.html.erb:1]: Same here

```suggestion
<% add_decidim_page_title(translated_attribute(election.title)) %>
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/elections/new.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/elections/new.html.erb:1]: Same here
```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/elections/publish_page.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/elections/publish_page.html.erb:1]: Same here
```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/questions/edit.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/questions/edit.html.erb:1]: Same here
```suggestion
<% add_decidim_page_title(translated_attribute(question.title)) %>
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/questions/index.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/questions/index.html.erb:1]: Same here
```suggestion
<% add_decidim_page_title(translated_attribute(election.title)) %>
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/questions/new.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/questions/new.html.erb:1]:
```suggestion
<% add_decidim_page_title(t(".title")) %>

```

File: app/views/decidim/vocdoni/admin/shared/_election_title.html.erb
Comment (andreslucena) [inline app/views/decidim/vocdoni/admin/shared/_election_title.html.erb:1]:
```suggestion
<% add_decidim_page_title(translated_attribute(election.title)) %>

```

File: app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb
Comment (andreslucena) [inline app/helpers/decidim/vocdoni/admin/steps_wizard_helper.rb:24]: As we have some duplicated logic with the permission class, and as these two features (the permission check and the tab_link_class helper) are really two sides of the same coin, how about if we extract these methods to the Election model? What I have in mind is something like this: 
```suggestion
        def tab_link_class(tab_name, election)
          case tab_name
          when "questions"
            election.election.ready_for_questions_form? ? "" : "disabled"
          when "census"
            election.ready_for_census_form?  ? "" : "disabled"
          when "results"
            election.ready_for_results_form?  ? "" : "disabled"
          when "publish"
            election.ready_for_publish_form?  ? "" : "disabled"
```

And then in `app/models/decidim/vocdoni/election.rb`:


```ruby

def ready_for_questions_form?
  present?
end 

def ready_for_census_form?
  ...
end 
``` 

By reusing the logic on both of these places we only need to change it on one place in the future when/if the conditions need to change, and it's also easier to test and read.

**Review by andreslucena — CHANGES_REQUESTED** [review]
A couple more changes in my side, can you review them please? Thanks!

File: app/permissions/decidim/vocdoni/admin/permissions.rb
Comment (andreslucena) [inline app/permissions/decidim/vocdoni/admin/permissions.rb:72]: We should implement the new meths here: 
```suggestion
          toggle_allow(election.ready_for_questions_form?)
```

File: app/permissions/decidim/vocdoni/admin/permissions.rb
Comment (andreslucena) [inline app/permissions/decidim/vocdoni/admin/permissions.rb:76]: And here: 
```suggestion
          toggle_allow(election.ready_for_calendar_form?)
```

File: app/permissions/decidim/vocdoni/admin/permissions.rb
Comment (andreslucena) [inline app/permissions/decidim/vocdoni/admin/permissions.rb:82]: And here: 
```suggestion
          toggle_allow(election.ready_for_publish_form?)
```

File: app/permissions/decidim/vocdoni/admin/permissions.rb
Comment (andreslucena) [inline app/permissions/decidim/vocdoni/admin/permissions.rb:64]: With all the new changes, I think that we could get rid of this method
```suggestion
```

File: app/permissions/decidim/vocdoni/admin/permissions.rb
Comment (andreslucena) [inline app/permissions/decidim/vocdoni/admin/permissions.rb:88]: And here:
```suggestion
          toggle_allow(election.ready_for_census_form?)
```

File: app/views/decidim/vocdoni/admin/answers/edit.html.erb
Comment (andreslucena) [reply app/views/decidim/vocdoni/admin/answers/edit.html.erb:1]: I think these changes weren't applied, can you check it?

File: app/models/decidim/vocdoni/election.rb
Comment (andreslucena) [inline app/models/decidim/vocdoni/election.rb:79]: I think this method can be private, as it's only used by other methods of this class. If that's correct, let's make it private explicitly, please 😄

**Review by andreslucena — CHANGES_REQUESTED** [review]
I have a few more changes.

Also I've seen that there were some "resolved" comments that weren't actually applied. There's a couple features that aren't clear in the "Review" functionality of GitHub that are good to know: 

1. You can apply the suggestions directly through the Web UI, so you don't need to apply them manually (ie copying and pasting). 
2. If you apply a change on the git command line, the comments review that apply to a given section will be marked resolved directly (I understand that you probably applied those changes locally but haven't pushed yet?) 

In either case, can you check the new comments please? Thanks

**Comment by antopalidi** [comment]
@andreslucena I'm sorry, I didn't push all the changes to the github yesterday. Now I'll double-check everything, fix the tests, and add new ones. And I'll text you when it's ready.

**Comment by antopalidi** [comment]
@andreslucena all done

**Review by andreslucena — COMMENTED** [review]
LGTM, 👍🏽

**Review by andreslucena — APPROVED** [review]

**Comment by andreslucena** [comment]
@microstudi you have requested changes and didn't give an approved review, so we can't merge for now: 

image

Can you give it a last look and approve it if everything is OK? Also feel free to merge it if that's the case, for me this is all great!

**Review by microstudi — APPROVED** [review]
Amazing work @antopalidi. Thanks for the thorough review @andreslucena !

**Pull request merged**
